driver/src/driver/scheduler/evm.rs

```diff
@@ -0,0 +1,279 @@
+//! batch duration information directly from the EVM instead of system time.
+
+use super::{AuctionTimingConfiguration, Scheduler, BATCH_DURATION};
+use crate::contracts::stablex_contract::StableXContract;
+use crate::driver::stablex_driver::{DriverResult, StableXDriver};
+use anyhow::Result;
+use log::{debug, error, info, warn};
+use std::thread;
+use std::time::Duration;
+
+/// The amount of time the scheduler should wait between polling.
+const POLL_TIMEOUT: Duration = Duration::from_secs(5);
+
+/// An EVM-based scheduler for the exchange driver.
+pub struct EvmScheduler<'a> {
+    exchange: &'a dyn StableXContract,
+    driver: &'a dyn StableXDriver,
+    config: AuctionTimingConfiguration,
+    last_batch: Option<u32>,
+}
+
+impl<'a> EvmScheduler<'a> {
+    /// Creates a new EVM-based scheduler from a configuration.
+    pub fn new(
+        exchange: &'a dyn StableXContract,
+        driver: &'a dyn StableXDriver,
+        config: AuctionTimingConfiguration,
+    ) -> Self {
+        EvmScheduler {
+            driver,
+            exchange,
+            config,
+            last_batch: None,
+        }
+    }
+
+    /// Creates a new scheduler with the default configuration.
+    #[cfg(test)]
+    pub fn with_defaults(exchange: &'a dyn StableXContract, driver: &'a dyn StableXDriver) -> Self {
+        EvmScheduler::new(exchange, driver, AuctionTimingConfiguration::default())
+    }
+
+    /// Runs the scheduler for a single iteration.
+    fn step(&mut self) -> Result<()> {
+        let batch_id = self.current_solving_batch()?;
+        if self
+            .last_batch
+            .map(|last_batch| batch_id <= last_batch)
+            .unwrap_or_default()
+        {
+            debug!("skipping already processed batch {}", batch_id);
+            return Ok(());
+        }
+
+        let time_remaining = self.exchange.get_current_auction_remaining_time()?;
+        // NOTE: We need to take into account the asynchronous nature of web3
+        //   and handle the case where we query the batch information right on
+        //   a batch border and the following happens:
+        //     - query batch ID and get `N`
+        //     - block gets mined, batch ID becomes `N+1`
+        //     - query time remaining and get around `300s`
+        //   In order to work around this, we just re-query the batch ID after
+        //   getting the time in the batch to make sure we are using the correct
+        //   batch. If they don't match, we just return to restart the run-loop.
+        let verify_batch_id = self.current_solving_batch()?;
+        if batch_id != verify_batch_id {
+            info!(
+                "batch ID changed during run loop ({} -> {}); retrying",
+                batch_id, verify_batch_id,
+            );
+            return Ok(());
+        }
+
+        let current_batch_time = BATCH_DURATION - time_remaining;
+        if current_batch_time > self.config.target_start_solve_time {
```

Comment:
Shouldn't this be latest_solve_attempt_time? We have seen in the past that we might not mine a block within the first 30s of a batch.

Reply:
:man_facepalming: